Firestore/core/src/firebase/firestore/remote/grpc_stream.cc

```diff
@@ -209,10 +250,22 @@ void GrpcStream::OnRead(const grpc::ByteBuffer& message) {
 
 void GrpcStream::OnWrite() {
   if (observer_) {
-    StreamWrite* maybe_write = buffered_writer_.DequeueNextWrite();
-    if (maybe_write) {
-      Execute(maybe_write);
+    absl::optional<grpc::ByteBuffer> maybe_write =
+        buffered_writer_.DequeueNextWrite();
+    if (!maybe_write) {
+      return;
     }
+    auto* completion = NewCompletion(
+        [this](bool ok, const GrpcStreamCompletion& completion) {
+          if (ok) {
```

Comment:
Is there any way for us to avoid repeating all this drudgery in all the completion handlers?

They all look this same:
```c++
if (ok) {
  // successful completion
} else {
  OnOperationFailed();
}
RemoveOperation(&completion);
```
Also, for some of these types we're converting back and forth between reference and pointer types. Could we just pass GrpcCompletion objects by pointer/const pointer?

If NewCompletion took a std::function<void(const GrpcCompletion*)> could it wrap that in function that did all the ok checking and removal?

Alternatively, could we have a method like this?

```c++
bool Proceed(bool ok, const GrpcCompletion* completion) {
  if (!ok) {
    OnOperationFailed();
  }
  RemoveOperation(completion);
  return ok;
}
```

Then callers would look like:

```c++
auto* completion = NewCompletion(
    [this](bool ok, const GrpcCompletion* completion) {
      if (Proceed(ok, completion)) {
        OnWrite();
      }
    });
```

I wish there were some way to avoid passing/naming ok and completion though.

Reply:
Sure, I didn't want to polish it before making sure you like the approach.

I prefer wrapping in a function (it results in a little less boilerplate). I can't think of a good way to avoid passing `completion` other than lambda capture (which would necessitate two-step initialization, which doesn't seem like a great trade-off).